lib/syscall_shim/include/uk/syscall_exittab.h

```diff
@@ -0,0 +1,104 @@
+extern const struct uk_syscall_exittab_entry uk_syscall_exittab_start[];
+extern const struct uk_syscall_exittab_entry uk_syscall_exittab_end;
+
+extern __thread unsigned long uk_syscall_nested_depth;
+
+/**
+ * Helper macro for iterating over system call exit functions.
+ * Please note that the table may contain NULL pointer entries and that a
+ * Unikraft system call exit function is only called during the
+ * exiting of the first called system call in the callchain (i.e. if a system
+ * call calls another system call in Unikraft context (although it shouldn't)
+ * the routines in this table is iterated upon only when exiting the first
+ * called system call).
+ *
+ * @param itr
+ *   Iterator variable (struct uk_syscall_exittab_entry *) which points to the
+ *   individual table entries during iteration
+ * @param syscall_exittab_start
+ *   Start address of table (type: const struct uk_syscall_exittab_entry[])
+ * @param syscall_exittab_end
+ *   End address of table (type: const struct uk_syscall_exittab_entry)
+ */
+#define uk_syscall_exittab_foreach(itr, tab_start, tab_end)		\
+	for ((itr) = DECONST(struct uk_syscall_exittab_entry *, tab_start);\
+	     (itr) < &(tab_end);					\
+	     (itr)++)
+
+/**
+ * Register a Unikraft system call exit function that is called during the
+ * exiting of the first called system call in the callchain. For example,
+ * if a binary system call is called and during its handling some other
+ * native system calls (just the function calls, no `syscall`/`svc`/etc
+ * instructions involved) are called during the handling of the binary system
+ * call, then this entry is executed only once, during the exit of the binary
+ * system call's exit. Likewise, if in the current callchain there are only
+ * native system calls, then the entry will get executed during the exit of the
+ * first called system call.
+ *
+ * @param fn
+ *   System call exit function to be called
+ * @param prio
+ *   Priority level (0 (earliest) to 9 (latest)), must be a constant.
+ *   Use the UK_PRIO_AFTER()/UK_PRIO_BEFORE() helper macro for computing
+ *   priority dependencies.
+ *   Note: Any other value for level will be ignored
+ */
+#define __UK_SYSCALL_EXITTAB_ENTRY(fn, prio)				\
+	static const struct uk_syscall_exittab_entry			\
+	__used __section(".uk_syscall_exittab" #prio) __align(8)	\
+		__uk_syscall_exittab ## prio ## _ ## fn = {		\
+		.func = (fn),						\
+	}
+
+#define _UK_SYSCALL_EXITTAB(fn, prio)					\
+	__UK_SYSCALL_EXITTAB_ENTRY(fn, prio)
+
+#define uk_syscall_exittab_prio(fn, prio)				\
+	_UK_SYSCALL_EXITTAB(fn, prio)
+
+static inline void uk_syscall_exittab_run(struct uk_syscall_exit_ctx *sctx)
+{
+	struct uk_syscall_exittab_entry *entry;
+
+	UK_ASSERT(sctx);
+
+	uk_pr_debug("Sycall exit table @ %p - %p\n",
+		    &uk_syscall_exittab_start[0], &uk_syscall_exittab_end);
+
+	uk_syscall_exittab_foreach(entry,
+				   uk_syscall_exittab_start,
+				   uk_syscall_exittab_end) {
+		UK_ASSERT(entry);
+
+		if (!entry->func)
```

Comment:
I would recommend that you add a comment on why you're returning here, or even better a comment in the declaration of `struct uk_syscall_*_ctx` that `execenv` can be `NULL` (and under which circumstances), as all handlers need to be aware of that.